tsdb/importer/importer.go

```diff
@@ -0,0 +1,602 @@
+)
+
+type blockTimestampPair struct {
+	start, end int64
+}
+
+type newBlockMeta struct {
+	index      int
+	count      int
+	mint, maxt int64
+	isAligned  bool
+	dir        string
+	children   []*newBlockMeta
+}
+
+// Content Type for the Open Metrics Parser.
+// Needed to init the text parser.
+const contentType = "application/openmetrics-text;"
+
+// ImportFromFile imports data from a file formatted according to the Open Metrics format,
+// converts it into block(s), and places the newly created block(s) in the
+// TSDB DB directory, where it is treated like any other block.
+func ImportFromFile(
+	r io.Reader,
+	dbPath string,
+	maxSamplesInMemory int,
+	maxBlockChildren int,
+	logger log.Logger,
+) error {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	level.Debug(logger).Log("msg", "creating temp directory")
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDbDir)
+
+	level.Debug(logger).Log("msg", "reading existing block metadata to infer time ranges")
+	blockMetas, err := constructBlockMetadata(dbPath)
+	if err != nil {
+		return err
+	}
+
+	level.Info(logger).Log("msg", "importing input data and aligning with existing DB")
+	if err = writeSamples(r, tmpDbDir, blockMetas, maxSamplesInMemory, maxBlockChildren, logger); err != nil {
+		return err
+	}
+
+	level.Info(logger).Log("msg", "merging fully overlapping blocks")
+	if err = mergeBlocks(tmpDbDir, blockMetas, logger); err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "copying newly created blocks from temp location to actual DB location")
+	if err = fileutil.CopyDirs(tmpDbDir, dbPath); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// writeSamples parses each metric sample, and writes the samples to the correspondingly aligned block
+// to the given directory.
+func writeSamples(
+	r io.Reader,
+	dir string,
+	metas []*newBlockMeta,
+	maxSamplesInMemory int,
+	maxBlockChildren int,
+	logger log.Logger,
+) error {
+	// First block represents everything before DB start.
+	// Last block represents everything after DB end.
+	dbMint, dbMaxt := metas[0].maxt, metas[len(metas)-1].mint
+
+	blocks := make([][]*tsdb.MetricSample, len(metas))
+	currentPassCount := 0
+
+	// TODO: Maybe use bufio.Reader.Readline instead?
+	// https://stackoverflow.com/questions/21124327/how-to-read-a-text-file-line-by-line-in-go-when-some-lines-are-long-enough-to-ca
+	// Use a streaming approach to avoid loading too much data at once.
+	scanner := bufio.NewScanner(r)
+	buf := new(bytes.Buffer)
+	scanner.Split(sampleStreamer(buf))
+
+	for scanner.Scan() {
+		if currentPassCount == 0 {
+			blocks = make([][]*tsdb.MetricSample, len(metas))
+		}
+
+		encSample := scanner.Bytes()
+		decBuf := bytes.NewBuffer(encSample)
+		sample := tsdb.MetricSample{}
+		if err := gob.NewDecoder(decBuf).Decode(&sample); err != nil {
+			level.Error(logger).Log("msg", "failed to decode current entry returned by file scanner", "err", err)
```

Comment:
Let's avoid things like this, either log error or return, let's not handle errors twice (: